sdk/core/core-rest-pipeline/src/retryStrategies/failoverRetryStrategy.ts

```diff
@@ -0,0 +1,163 @@
+interface FailoverHosts {
+  index: number;
+  hostStates: FailoverHostState[];
+}
+
+/**
+ * Default behavior for generating hosts for failover.
+ *
+ * If the request method is GET, HEAD, or OPTIONS, it cycles through `readHosts`. Otherwise, it cycles through `writeHosts`.
+ */
+export function readWriteFailoverHostDelegate(options: {
+  readHosts?: string[];
+  writeHosts?: string[];
+}): FailoverHostDelegate {
+  const { readHosts, writeHosts } = options;
+  const readHostStateList: FailoverHosts = {
+    index: 0,
+    hostStates: (readHosts ?? []).map((url) => {
+      const host = new URL(url).host;
+      return { host, retryCount: 0 };
+    }),
+  };
+  const writeHostStateList: FailoverHosts = {
+    index: 0,
+    hostStates: (writeHosts ?? []).map((url) => {
+      const host = new URL(url).host;
+      return { host, retryCount: 0 };
+    }),
+  };
+
+  return function* (retryState, factoryOptions) {
+    // Optimization for the case where no retry hosts are specified
+    if (!readHostStateList.hostStates.length && !writeHostStateList.hostStates.length) {
+      return;
+    }
+
+    while (true) {
+      const response = retryState.response;
+      const retryDelayInMs = factoryOptions.retryDelayInMs ?? DEFAULT_CLIENT_RETRY_INTERVAL;
+      const maxRetryDelayInMs =
+        factoryOptions.maxRetryDelayInMs ?? DEFAULT_CLIENT_MAX_RETRY_INTERVAL;
+      if (!response) {
+        return;
+      }
+
+      // Select correct list of hosts for this request
+      const isSafeMethod = ["GET", "HEAD", "OPTIONS"].includes(response.request.method);
+      const currentStateList = isSafeMethod ? readHostStateList : writeHostStateList;
+
+      if (currentStateList.hostStates.length === 0) {
+        yield undefined;
+        continue;
+      }
+      const hostState = currentStateList.hostStates[currentStateList.index];
+
+      const requestFireTime = Math.max(hostState.retryAt?.valueOf() ?? -Infinity, Date.now());
+
+      retryState = yield hostState;
+
+      const exponentialDelay = exponentialDelayInMs(
+        hostState.retryCount,
+        retryDelayInMs,
+        maxRetryDelayInMs
+      );
+      const nextRetryAt = new Date(requestFireTime + exponentialDelay);
+      hostState.retryAt = nextRetryAt;
+
+      currentStateList.index = (currentStateList.index + 1) % currentStateList.hostStates.length;
+      hostState.retryCount++;
+    }
+  };
+}
+
+/**
+ *
+ * @param options - Lists of fallback endpoints for read/write operations
+ * @returns An exponential retry strategy with support for multiple endpoints
+ */
+export function failoverRetryStrategy(options: DefaultRetryPolicyOptions): RetryStrategy {
+  const generatorMap = new WeakMap<PipelineRequest, ReturnType<FailoverHostDelegate>>();
+  return {
+    name: "failoverRetryStrategy",
+    retry(state) {
+      const response = state.response;
+      // Only fallback to another host on a 5xx response
+      if (!(response && 500 <= response.status && response.status < 600)) {
```

Comment:
I wonder if we should also retry on system errors where the host is unreachable?

Reply:
`ETIMEDOUT`, `ECONNREFUSED`, and `ENOENT` are handled in this retry strategy now. Though I can't quite figure out which syscall returns `ENOENT`.

Reply:
That one is DNS lookup failure IIRC

Reply:
It took 5 seconds of glancing at the source of libc `getaddrinfo` for my eyes to glaze over, so I'll take your word for it. It can stay here until the day it breaks. Which is hopefully never.